hw/xbox/xbox.c

```diff
@@ -253,7 +253,53 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
 /* PC hardware initialisation */
 static void xbox_init(MachineState *machine)
 {
-    xbox_init_common(machine, default_eeprom, NULL, NULL);
+    char *filename;
+    int fd;
+    int rc;
+    int eeprom_file_size;
+    const int eeprom_size = 256;
+    uint8_t *eeprom_data = g_malloc(eeprom_size);
+
+    const char *eeprom_file = object_property_get_str(qdev_get_machine(),
+                                                      "eeprom", NULL);
+    if ((eeprom_file != NULL) && *eeprom_file) {
+        filename = qemu_find_file(QEMU_FILE_TYPE_BIOS, eeprom_file);
+        assert(filename);
+
+        eeprom_file_size = get_image_size(filename);
+        if (eeprom_size != eeprom_file_size) {
+            fprintf(stderr,
+                    "qemu: EEPROM file size != 256 bytes. (Is %i bytes)\n",
+                    eeprom_file_size);
+            g_free(filename);
+            exit(1);
+            return;
+        }
+
+        fd = open(filename, O_BINARY);
+        if (fd < 0) {
+            fprintf(stderr, "qemu: EEPROM file '%s' could not be opened.\n", filename);
+            g_free(filename);
+            exit(1);
+            return;
+        }
+
+        rc = read(fd, eeprom_data, eeprom_size);
+        if (rc != eeprom_size) {
+            fprintf(stderr, "qemu: Could not read the full EEPROM file.\n");
+            close(fd);
+            g_free(filename);
+            exit(1);
+            return;
+        }
+
+        close(fd);
+        g_free(filename);
+    } else {
+        memcpy(eeprom_data, default_eeprom, eeprom_size);
+    }
```

Comment:
I think the complexity of this code section warrants putting it in a separate function (~`load_eeprom` or something)

Reply:
Agree (what I was hinting at with above comment too)